appcelerator/titanium_mobile
# refactor(android): build with gradle

**JIRA:**
- https://jira.appcelerator.org/browse/TIMOB-19300
- https://jira.appcelerator.org/browse/TIMOB-24517
- https://jira.appcelerator.org/browse/TIMOB-25240
- https://jira.appcelerator.org/browse/TIMOB-27325

**Hyperloop Warning:**
This PR will temporarily break hyperloop. Once this PR has been merged, we can then rebuild hyperloop module with gradle via the following private hyperloop PR...
https://github.com/appcelerator/hyperloop.next/pull/329.

**App Build Time Improvements:**
Benchmarked with [kitchensink-v2](https://github.com/appcelerator/kitchensink-v2) on a 15" MacBook Pro 2015 with a solid state drive.
```
Non-encrypted emulator builds:
- Clean Build: 52s -> 20s  (2.6x faster)
- Incremental: 18s -> 7s   (2.6x faster)

Encrypted device/production builds:
- Clean Build: 53s -> 21s  (2.5x faster)
- Incremental: 50s -> 14s  (3.6x faster)

Notes:
- Benchmarks exclude gradle daemon startup of ~5s which is a one time hit.
- Titanium did not support encrypted incremental builds before. This is new.
- App gradle builds will be faster once we rebuild modules as AARs.
```

**Summary of Changes:**
- Replaced "ant" tool usage with "gradle" for SDK, app, and module builds.
- Updated min Android build-tools requirement from 25.0.0 to 28.0.3.
- Titanium SDK Library:
  * Now built as a single AAR instead of individual JARs.
  * Can now easily open `titanium_mobile/android` directory in Android Studio.
  * Replaced `titanium_mobile/android` project `TitaniumTest` with `app`.
  * Supports building debug version of C/C++ `.so` libraries for asserts and debug logging.
  * Removed all Google JARs and res files and replaced with gradle/maven references.
  * Transpile/polyfill/rollup/snapshot of "ti.main.js" common folder is now done as a pre-build step of "titanium" library. Supports incremental builds when JS files change.
  * Jenkins/node-scons builds will now fail if unable to acquire V8 snapshot.
- Apps:
  * Significantly improved build times. Now supports incremental device/production builds.
  * Now automatically digitally signs APK with signature scheme v2.
  * Now merges module/app "AndroidManifest.xml" files via gradle.
  * Now supports "AndroidManifest.xml" merge `tools:` attributes. See [link](https://developer.android.com/studio/build/manifest-merge)
  * Added support for `./platform/android/build.gradle` file for defining dependencies.
  * Build no longer post-fixes hashes to app's `res/drawable` image files. Now stored as-is.
  * Building for "development" or "test" now builds a debug version. Can be used to debug modules and SDK via Android Studio.
  * Added new build hook event `build.android.requestResourcesDirPaths`. To be used by plugins/modules (such as hyperloop) to add additional "Resources" files to app build.
- Modules:
  * Are now built as AARs and installed as local maven repos for version management.
  * Modules built with 9.0.0 are not backward compatible. Must set "minsdk" to 9.0.0.
  * Now supports "Resources" folder whose files are copied into app as-is (like iOS).
  * Added support for `./platform/android/AndroidManifest.xml` file.
- Dropped support for appc config option `android.mergeCustomAndroidManifest`.
  * Must use  XML `tools:remove` or `tools:removeAll` attributes instead which is Google's equivalent feature.

---
**KitchenSink Test:**
_Test the below on Mac and Windows._
1. Download the [kitchensink-v2](https://github.com/appcelerator/kitchensink-v2) app.
2. Change its "tiapp.xml" file to reference this PR's SDK version.
3. Build and run to the Android emulator.
4. Verify app's top and bottom bars are RED. _(Indicates custom theme XML file works.)_
5. Tap through all of the app's UI, verify that everything works. _Particular Maps._
6. Build and run to the emulator again. (Tests increment build.)
7. Tap through all of the app's UI again.
8. Run it on emulator via Appc Studio debugger. Verify break-points work.
9. Run it on emulator via Appc Studio Live-View feature. Verify JS changes re-launches app.
10. Build and run to an Android device. (Tests encrypted JS builds.)
11. Tap through all of the app's UI again.
12. Build and run to the device again. (Tests incremental encrypted builds.)
13. Tap through all of the app's UI again.
14. Run it on device via Appc Studio debugger. Verify break-points work.
15. Run it on device via Appc Studio Live-View feature. Verify JS changes re-launches app.
16. Build app for production. (This requires a keystore file for signing app.)
17. Uninstall app from device/emultor.
18. Install and run production app on device/emulator.
19. Tap through all of the app's UI again.

---
**APK Signature Scheme v2 Test:**
1. Following the test procedure attached to [TIMOB-24517](https://jira.appcelerator.org/browse/TIMOB-24517).
2. Verify command line tool returns `true` for both v1 and v2 signatures.
_(Note that v3 should be `false`. Not currently supported.)_

---
**"AndroidManifest.xml" Error Handling Test:**
1. Follow test procedure attached to [TIMOB-27325](https://jira.appcelerator.org/browse/TIMOB-27325).
2. Build the app.
3. Verify that app build fails and it logs what's wrong with the XML. _(This is a good thing.)_

---
**Custom "build.gradle" Test:**
1. Create a Classic app.
2. Copy the below `app.js` to the project.
3. Copy the below `build.gradle` file to project's `./platform/android` folder.
4. Build and run on Android.
5. Verify black "play" button is shown onscreen. _(Comes from gradle file's referenced library.)_
6. In build log, verify you see message: `@@@ [build.gradle] JavaCompile.doFirst`

./Resources/app.js
```javascript
var window = Ti.UI.createWindow({
	backgroundColor: "white",
});
window.add(Ti.UI.createImageView({
	image: "android.resource://" + Ti.App.id + "/drawable/exo_notification_small_icon"
}));
window.open();
```

./platform/android/build.gradle
```groovy
repositories {
	google()
	jcenter()
}
tasks.withType(JavaCompile) {
	doFirst {
		println('@@@ [build.gradle] JavaCompile.doFirst')
	}
}
dependencies {
	implementation 'com.google.android.exoplayer:exoplayer-ui:2.8.0'
}
```
